src/main/java/org/jenkinsci/plugins/workflow/log/FileLogStorage.java

```diff
@@ -0,0 +1,276 @@
+                                    LOGGER.warning("Ignoring corrupt index file " + index);
+                                }
+                                lastId = space == -1 ? null : line.substring(space + 1);
+                            }
+                            if (pos == lastTransition) {
+                                if (hadLastId) {
+                                    w.write(LogStorage.endStep());
+                                }
+                                hadLastId = lastId != null;
+                                if (lastId != null) {
+                                    w.write(LogStorage.startStep(lastId));
+                                }
+                            }
+                            super.write(b);
+                            pos++;
+                        }
+                        @Override public void flush() throws IOException {
+                            if (lastId != null) {
+                                w.write(LogStorage.endStep());
+                            }
+                            super.flush();
+                        }
+                    });
+                    ConsoleAnnotators.setAnnotator(caos.getConsoleAnnotator());
+                    return r;
+                }
+            }
+        };
+    }
+
+    @Override public AnnotatedLargeText<FlowNode> stepLog(FlowNode node, boolean complete) {
+        String id = node.getId();
+        try (ByteBuffer buf = new ByteBuffer();
+             RandomAccessFile raf = new RandomAccessFile(log, "r");
+             BufferedReader indexBR = index.isFile() ? Files.newBufferedReader(index.toPath(), StandardCharsets.UTF_8) : new BufferedReader(new NullReader(0))) {
+            String line;
+            long pos = -1; // -1 if not currently in this node, start position if we are
+            while ((line = indexBR.readLine()) != null) {
+                int space = line.indexOf(' ');
+                long lastTransition = -1;
+                try {
+                    lastTransition = Long.parseLong(space == -1 ? line : line.substring(0, space));
+                } catch (NumberFormatException x) {
+                    break; // corrupt index file; forget it
+                }
+                if (pos == -1) {
+                    if (space != -1 && line.substring(space + 1).equals(id)) {
+                        pos = lastTransition;
+                    }
+                } else {
+                    raf.seek(pos);
+                    if (lastTransition > pos + Integer.MAX_VALUE) {
+                        throw new IOException("Cannot read more than 2Gib at a time"); // ByteBuffer does not support it anyway
+                    }
+                    // TODO can probably be done a bit more efficiently with FileChannel methods
+                    byte[] data = new byte[(int) (lastTransition - pos)];
+                    raf.readFully(data);
+                    buf.write(data);
```

Comment:
🐜 Could we not write this a chunk at a time or stream it?  This is going to generate a *lot* of GC pressure due to excess array allocation.

Reply:
Sort of beside the point when the whole `buf` is memory-allocated. Which is hard to get around in current core versions since the only `[Annotated]LargeText` constructors take a file (useless) or a memory buffer. To write this properly, we would likely need for the [`Source`](https://github.com/stapler/stapler/blob/cafcb607e9e957bc8f57782eab3b0568a2ab3f42/core/src/main/java/org/kohsuke/stapler/framework/io/LargeText.java#L60-L67) and [`Session`](https://github.com/stapler/stapler/blob/cafcb607e9e957bc8f57782eab3b0568a2ab3f42/core/src/main/java/org/kohsuke/stapler/framework/io/LargeText.java#L381-L390) interfaces to be made `public`.

Reply:
Answering my own question after digging in a bit more: no, because our `ByteBuffer` cannot be pre-sized for our content since its `ensureCapacity` method is private.

For those who can't follow the code: essentially what we're doing it allocating an array containing all the log content and then copying it into the Jenkins-internal `ByteBuffer` implement (`org.kohsuke.stapler.framework.io.ByteBuffer`) which will dynamically resize to fit the content.  Essentially we're double-allocating.  

However the ByteBuffer will resize to *either* double *or* fit content (whichever is larger).   Which means that if we copied a chunk at a time, we'd likely end up triggering multiple backing-array resize operations for the `ByteBuffer.`

Kind of a wash in that case, without a way to more effectively size the buffer, which would require upstream changes to Stapler (which can be done later).

Reply:
AFAICT the `FileChannel` comment -- which is another clever option -- won't work because our `ByteBuffer` (drumroll please) _does not_ extend off of `java.nio.ByteBuffer`... instead it extends `OutputStream`  😆 

So... I smell some room for core improvements in LargeText/AnnotatedLargeText that reduce our tendency to throw around (needless) byte array allocations.  Might help with some of the humongous object allocations I've seen spiking our GC logs.

Reply:
Fair enough -- nothing to do here at the moment but I'd like to leave the conversation marked unresolved to nudge us about future enhancements to do.